Validate instance config on a per-instance-type basis

my5t3ry opened the pull request:
Resolves: #8912

Signed-off-by: Sascha Bast <sash@mischkonsum.org>

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -125,8 +125,43 @@ func exclusiveConfigKeys(key1 string, key2 string, config map[string]string) (va
+	"security.protection.shift",
+	"security.syscalls",
+	"security.privileged",
+	"limits.memory.swap.priority"}
```
Trailing `}` needs to be on newline (did you run this through `gofmt` btw?).

my5t3ry replied:
I usually run gofmt -s -w ./.. before I commit something. I ran it again for this file, but it seems that it does not realign this bracket.

tomponline replied:
@my5t3ry ah ok in that case its just our own project specific convention then :)

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -125,8 +125,43 @@ func exclusiveConfigKeys(key1 string, key2 string, config map[string]string) (va
 	return "", false, nil
 }
 
+var containerSpecificConfigKeys = []string{
```
I think I would have preferred using a `map[instancetype.Type][]string` and then stored both container and VM specific keys in the same map, that way they can be looked up by the provided `instanceType` argument in `ValidConfig`.

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -140,6 +175,18 @@ func ValidConfig(sysOS *sys.OS, config map[string]string, profile bool, expanded
 		}
 
+		if !profile {
+			if instanceType == instancetype.VM {
```
Lets simplify this bit based on the change suggested in https://github.com/lxc/lxd/pull/8970/files#r662828712

You can just do:

```go
for specificInstType, specificKeys := range specificInstTypeConfigKeys {
    if specificInstType == instanceType {
        continue
    }

    if shared.StringInSlice(k, specificKeys) {
         return fmt.Errorf("Config key %q is not allowed for instance type %s", k, instanceType)
    }
}

```

tomponline replied:
Also in `isContainerSpecificConfigKey` I'm not sure using ` strings.Contains` is a good idea as it may mismatch partial string keys. Was this intentionally added? Rather than using exact match as in my example above?

my5t3ry replied:
i used strings.Contains because some key namespaces are instance type exclusive like "limits.hugepage" for Containers, and i could check on the whole namespace instead of key.

tomponline replied:
@my5t3ry ah I see so you're intention was to do a prefix check. Using `strings.Contains` is still not desirable because it may match a future part of a key unintentionally. `strings.HasPrefix()` would be better then in that case. Although we would need to define a different validation function for each key in the list because some should be exact match and not allow prefixes, and others are prefix based. Although in this case limits.hugepages.* are all well defined keys and so shouldn't be using prefix match anyway.

tomponline reviewed (changes requested):
Thanks, few minor changes there.

How did you come to build the container and VM specific config key lists by the way?

Although thinking about this more, I think I may be tempted to turn this on its head and define a single list of generic keys allowed across all instance types (including profiles), and then have a set of keys that are allowed on a per instance type basis (which are also allowed in profiles).

This way when we add a new key in the future we will have to remember to add it to either the generic list or the type specific list, whereas using a deny list style like now and by default allowing all other keys to be used with any instance type is asking for trouble in the future as people (myself included) tend to forget to add them to a specific list.

Also, if we had a list of generic keys, and type specific keys, then we could also validate that a profile doesn't use a mixture of keys across instance types, as although we can't be sure what instance type a profile will be used with, we can be sure that if a profile mixes instance type specific keys then the profile will be unusable for all instance types after this validation is applied.

tomponline commented:
jenkins: test this please

my5t3ry commented:
hi, thanks for your review. i got it and will try to implement your suggested approach at some point this weekend. Should I also add the validation for profiles (only generic and no mixed instance specific keys)? 
I checkd on why the testsuite failed and found this:
`EROR[07-02|09:10:15] Device stop validation failed for        instanceType=container instance=proxyTester project=default device=proxyDev type=proxy err="Invalid value for device option \"connect\": Not an IP address \"localhost\""
EROR[07-02|09:10:16] Device remove validation failed          instanceType=container instance=proxyTester project=default device=proxyDev type=proxy err="Invalid value for device option \"connect\": Not an IP address \"localhost\""`

I guess this is not related to my changes, or do i get this wrong? 

: )

tomponline commented:
@my5t3ry thanks! The only test failure for your PR was https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/3930/arch=amd64,backend=lvm,compiler=golang-1.15/console#console-section-8 and its not related to your change.

tomponline commented:
jenkins: test this please

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -126,7 +126,7 @@ func exclusiveConfigKeys(key1 string, key2 string, config map[string]string) (va
 
 // ValidConfig validates an instance's config.
-func ValidConfig(sysOS *sys.OS, config map[string]string, profile bool, expanded bool) error {
+func ValidConfig(sysOS *sys.OS, config map[string]string, profile bool, expanded bool, instanceType instancetype.Type) error {
```
@my5t3ry because you only provide `instanceType = instancetype.Any` when `profile = true`, then I think we can remove the `profile` argument from this function entirely, and just use `instanceType = instancetype.Any` to indicate profile validation (with an updated doc block comment explaining that). Otherwise we end up with two arguments that do very similar things which is confusing.

tomponline commented:
You have a golint error preventing tests running:

```
lxd/instance/instance_utils.go:161:25: error strings should not be capitalized or end with punctuation or a newline
```

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -140,6 +140,33 @@ func ValidConfig(sysOS *sys.OS, config map[string]string, profile bool, expanded
 		}
 
+		if !profile {
+			for specificInstType, specificKeys := range specificInstTypeConfigKeys {
```
@my5t3ry if I'm reading this right, for non-profile validation, this won't allow an instance to use the `instancetype.Any` profile keys (which should be allowed for any type of instance).

tomponline commented:
jenkins: test this please

my5t3ry commented:
it seems that "security.syscalls.deny" and "security.secureboot" are not instance specific like stated in https://lxd.readthedocs.io/en/latest/instances/. I get `
Config parsing error: Invalid expanded config: Config key "security.syscalls.deny" is not allowed for instance type virtual-machine
Press enter to open the editor again or ctrl+c to abort change
`
and
`
Config parsing error: Invalid expanded config: Config key "security.secureboot" is not allowed for instance type container
Press enter to open the editor again or ctrl+c to abort change
`

im moved them to instancetype.Any

tomponline commented:
That seems correct, they are instance type specific, `security.secureboot` is only for VMs and `security.syscalls.deny` is only for containers.

my5t3ry commented:
sry, i got confused because I added the config keys to a profile which was assigned to a vm and a container. I moved them back.

tomponline commented:
jenkins: test this please

my5t3ry commented:
there was still a bug with creating profils. it's fixed now.

btw, should I add tests to the testsuite for this?

tomponline reviewed (changes requested):
So I still think this would be better changed around (like I described the other day) so that rather than checking each key *isn't* one of the disallowed keys, instead checking that each key *is* one of the allowed keys for the instance type.

This way we always have to remember to update this list when adding new keys, whereas right now I can see us forgetting to add to this list in the future.

For example right now if a key is missing entirely from all the lists, then it would be allowed, rather than being denied.

However because there are some keys that are prefixes rather than specific keys you'll need to handle different styles of checks. I'm thinking it may be easier to extend the structure of `KnownInstanceConfigKeys` and turn it into a `map[instancetype.Type]map[string]func(value string)` and then split the existing validation definitions there by instance type. Then update the functions that use that variable to handle instance type being passed to it.

What do you think?

As for testing, yes I do think we should have some checks added to ensure that using a VM key on a container is prevented, and that using a mixture of instance type keys on a profile is prevented.

Thanks

my5t3ry commented:
thank you. sry, I totally missed the change around part.

i added a to test_basic_usage but have struggle passing the test. It works if I run them on the lxd i started locally with the changes but it seems that local changes aren't included in the lxd which is spawned by the test suite. Can you maybe give me a hint how make the testsuite run including the changes?

i  found one undocumented key: migration.stateful.
and one missing validator: security.syscalls.intercept.bpf which i added to to the container specific keys.

tomponline commented on lxd/instance/instance_utils.go:
```diff
@@ -1173,3 +1173,5 @@ func NextSnapshotName(s *state.State, inst Instance, defaultPattern string) (str
 	return pattern, nil
 }
+
+var specificInstTypeConfigKeys = map[instancetype.Type][]string{}
```
Looks like this var can go now.

tomponline commented on lxd/profiles_utils.go:
```diff
@@ -232,3 +232,21 @@ func getProfileInstancesInfo(cluster *db.Cluster, projectName string, profileNam
 	return instances, nil
 }
+
+func validateProfileConfig(d *Daemon, config map[string]string) error {
```
Why can't you do this as part of the existing `ValidConfig` function (just wrapped in an if for `instancetype.Any`)? Because with this change, you're now skipping a bunch of extra checks that occur in `ValidConfig` that are not in here? It would be good to keep instance validation (be it for profile or real instances) together.

my5t3ry replied:
I thougth because i's profile specific it should be placed in `profiles_utils.go`, and its calling `ValidConfig` at `validateProfileConfig` also. Should i move the proflie validation logic back to  `ValidConfig` ?

tomponline replied:
It would make sense, except that in doing that you're removed all this validation for profiles:

https://github.com/lxc/lxd/blob/6ad8779687c0987bfa7bbfe1ac3e5f547e1b7b93/lxd/instance/instance_utils.go#L149-L201

my5t3ry replied:
sry, i don"t get it. `ValidateConfig` is called when. The part you highlighted should be executed or am I wrong?

https://github.com/lxc/lxd/blob/5d14dbfc368443c3154c36f7c0bf8fa86aca29b2/lxd/profiles_utils.go#L236-L240

tomponline replied:
Oh I see you're calling instance.ValidConfig from inside `validateProfileConfig` I had thought it was instead of. OK that makes sense then to leave as is. As a general comment it would be easier to follow if you added some comments to your code to explain the logic being applied as I've found some of it non-trivial to reason about. Thanks

my5t3ry replied:
ok shure, no problem :)

tomponline commented on lxd/instance/instancetype/instance_type.go:
```diff
@@ -20,6 +20,13 @@ const (
 	VM = Type(1)
 )
 
+//GetInstanceTypes returns slice of existing Instance Type
```
"//GetInstanceTypes returns slice of existing Instance Types."

tomponline commented on test/suites/basic.sh:
```diff
@@ -542,4 +541,21 @@ test_basic_usage() {
+
+  # Test instance type specific config keys
+  lxc init testimage instance-type-specific-config-test-container
+  lxc init testimage instance-type-specific-config-test-vm
```
Sadly we cannot create VMs are part of the test suite.

my5t3ry replied:
ok, i removed the faulty vm test.

tomponline replied:
Let me know when the other issues (like go vet) are resolved and I'll re-run the tests. Thanks

tomponline commented on test/suites/basic.sh:
```diff
@@ -542,4 +541,21 @@ test_basic_usage() {
+  lxc config set instance-type-specific-config-test-container security.nesting true
+  lxc config set instance-type-specific-config-test-vm security.secureboot true
+  ! lxc config set instance-type-specific-config-test-vm security.nesting true || false
+  lxc profile delete instance-type-specific-config-test-container
```
You've not delete the instances you created above.

tomponline commented:
jenkins: test this please

tomponline commented:
> I found one undocumented key: migration.stateful.
> And one missing validator: security.syscalls.intercept.bpf which i added to to the container specific keys.

Ah good spot, please can you update the docs with these entries?

tomponline commented:
I've kicked off the jenkins tests now for you.

tomponline commented:
go vet failures:

```
vet: lxd/profiles.go:395:6: id declared but not used
```

tomponline commented:
jenkins: test this please

tomponline reviewed (approved):
Looks good. Just running the tests now to see if jenkins agrees.

Also, if you get time, it would be useful to us when reviewing and backporting changes into the LTS branches to have this PR split into multiple commits, rather than one large cross-file commit.

I tend to try and break the commits on logical changes or file boundaries.

tomponline reviewed (changes requested):
`vet: lxd/profiles.go:488:6: id declared but not used` still failing.

my5t3ry commented:
sry, i removed the unused id from `profilePut` but missed it @ `profilePatch`.

tomponline commented:
@my5t3ry a good example of breaking this change into multiple commits would be to have 1 for the removal of id from `doProfileUpdate` as it is unrelated to the main change.

tomponline commented:
jenkins: test this please

my5t3ry commented:
> @my5t3ry a good example of breaking this change into multiple logical commits would be to have 1 for the removal of `id` from `doProfileUpdate` as it is unrelated to the main change.

ok thank you, that's a helpful example. will try to structure my future commits this way.

tomponline commented:
Code style error:

```
ERROR: Functions must not start with an empty line: ./shared/instance.go-func ConfigKeyChecker(key string, instanceType instancetype.Type) (func(value string) error, error) {
```

tomponline commented:
jenkins: test this please

tomponline commented:
Looks like some jenkins errors:

https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/3972/arch=amd64,backend=dir,compiler=golang-1.13/console#console-section-12
```
 /lxc-ci/build/tmp.cACNqk6vjY/go/bin/lxc profile set default limits.cpu=2 --verbose
Error: Profiles with mixed instance specific configurations not allowed
```

my5t3ry commented:
> limits.cpu=2

this actuall was a bug. it failed if you tried to set a instancetype.Any config key after setting a instance type specific one. I fixed this one.

tomponline commented:
jenkins: test this please

my5t3ry commented:
@tomponline still doesn't do what it's expected to do. I have to look into it later today and will inform you as soon as it's fixed.

my5t3ry commented:
@tomponline i tried creating a profile like the profile which clould not be created at the test and it worked. The latest test you triggerd also seems to pass:

https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/3974/

Is it possible that it displayed 'Some checks failed' only because there where 2 test running at the same time?

tomponline commented:
jenkins: test this please

tomponline reviewed (approved):
Yep jenkins looks good now. Thanks

tomponline commented:
@my5t3ry I've asked @stgraber to have a quick look at this to check for anything else he thinks this PR should cover.

stgraber commented:
Hmm, so there appears to be one thing that's conceptually wrong, which is that a profile cannot include keys that are container or VM specific (or a mix of both).

Profiles aren't type-specific. I should be allowed to have a profile set `limits.memory.hugepages=true` while also setting `nvidia.runtime=true`. The former will only apply to VMs, the latter only to containers. That's fine.

There seems to be quite a bit of logic to figure out what the "type" of a profile is based on the keys it has set, and we just don't really need any of that. Profiles can contain a mix of both, that's fine.

Where we do need to validate the type is when setting configuration directly on an instance.

Other than that, we'll definitely want to split this up a bit. I'd probably recommend having one commit which just flips ValidConfig from the profile boolean over to taking the instance type. Then probably doing the functions in shared next as a separate commit. Then another commit to get the lxd side of things to adapt to that. Another for the CLI to similarly be updated (for lxc list). And then one last separate commit for the tests.

stgraber commented:
@tomponline makes sense?

tomponline commented:
> Hmm, so there appears to be one thing that's conceptually wrong, which is that a profile cannot include keys that are container or VM specific (or a mix of both).
> 
> Profiles aren't type-specific. I should be allowed to have a profile set `limits.memory.hugepages=true` while also setting `nvidia.runtime=true`. The former will only apply to VMs, the latter only to containers. That's fine.
> 
> There seems to be quite a bit of logic to figure out what the "type" of a profile is based on the keys it has set, and we just don't really need any of that. Profiles can contain a mix of both, that's fine.
> 
> Where we do need to validate the type is when setting configuration directly on an instance.
> 
> Other than that, we'll definitely want to split this up a bit. I'd probably recommend having one commit which just flips ValidConfig from the profile boolean over to taking the instance type. Then probably doing the functions in shared next as a separate commit. Then another commit to get the lxd side of things to adapt to that. Another for the CLI to similarly be updated (for lxc list). And then one last separate commit for the tests.

OK in that case I totally misunderstood what the desired behaviour should be. 

My understanding was that what ever config keys were in a profile would be applied to an instance when it is use (and so by extension if you mix keys of specific instance types in a profile then that profile would be unable to be applied to any single instance type). 

But if the application of profiles keys to an instance is expected to be more selective, and instead the instance just picks the ones it wants to use, then this does significantly change the idea behind it.

I think only this bit is blocking using mixed instance specific keys in a profile:

```go
	var profileKeyType = instancetype.Any
	for k := range config {
		_, t := shared.FindValidatorAndType(k)
		if profileKeyType != instancetype.Any && t != instancetype.Any && profileKeyType != t {
			return errors.New("Profiles with mixed instance specific configurations not allowed")
		}

		if t != instancetype.Any {
			profileKeyType = t
		}
	}
```

my5t3ry commented:
@tomponline i'm quite busy today but (hopfully) can finish this PR like @stgraber suggested tomorrow or does that you self assigned yourself to the PR mean that you'r taking over? 
thanks :)

tomponline commented:
@my5t3ry no thats fine for you to carry on when you get a chance. I assigned it to myself so I don't forget to review it.

my5t3ry commented:
@tomponline I am leaving for vacation  tomorrow for two weeks and unfortunately I will not be able to finish the PR before then. If it's not that urgent, I'll be happy to finish it as soon as I get back.

tomponline commented:
thanks @my5t3ry if I get chance before you're back I'll take a look.

stgraber commented:
Taking a look into this one, for now did a quick rebase.

tomponline commented:
> Taking a look into this one, for now did a quick rebase.

Thanks!
